lib/sudo-exec/src/pty.rs

```diff
@@ -51,11 +66,33 @@ impl PtyRelay {
         }
     }
 
-    fn wait_monitor(&mut self) -> io::Result<ExitReason> {
-        ExitReason::recv(&self.rx)
+    /// Read an event from the backchannel and return the event if it should break the event loop.
+    fn check_backchannel(&mut self) -> ControlFlow<ParentEvent> {
+        match self.backchannel.recv() {
+            // Not an actual error, we can retry later.
+            Err(err) if was_interrupted(&err) => {}
+            // Failed to read command status. This means that something is wrong with the socket
+            // and we should stop.
+            Err(err) => {
+                return ControlFlow::Break((&err).into());
+            }
+            Ok(event) => match event {
+                // Received the PID of the command. This means that the command is already
+                // executing.
+                ParentEvent::CommandPid(pid) => self.command_pid = pid,
+                // The command terminated or the monitor was not able to spawn it. We should stop
+                // either way.
+                ParentEvent::CommandExit(_)
+                | ParentEvent::CommandSignal(_)
+                | ParentEvent::IoError(_) => {
+                    return ControlFlow::Break(event);
+                }
+            },
+        }
+        ControlFlow::Continue(())
     }
 
-    fn relay_signal(&self, info: SignalInfo) {
+    fn relay_signal(&mut self, info: SignalInfo) {
```

Comment:
why did the mutability change here?

Reply:
I tried to do something else and failed, then I forgot to revert that change :sweat_smile: